src/symbolic_choice.ml

```diff
@@ -319,128 +217,198 @@ module Multicore = struct
+    let sat = Solver.check s thread.pc in
+    if sat then return () else stop
 
-  type 'a global_state =
-    { w : hold WQ.t (* work *)
-    ; r : ('a eval * Thread.t) WQ.t (* results *)
-    ; start_counter : Counter.t
-    }
+  let get_model symbol =
+    let* () = yield in
+    let* (S (solver_module, s)) = solver in
+    let module Solver = (val solver_module) in
+    let+ thread in
+    let sat = Solver.check s thread.pc in
+    if not sat then None
+    else begin
+      let model = Solver.model ~symbols:[ symbol ] s in
+      match model with
+      | None ->
+        failwith "Unreachable: The problem is sat so a model should exist"
+      | Some model -> begin
+        match Model.evaluate model symbol with
+        | None ->
+          failwith
+            "Unreachable: The model exists so this symbol should evaluate"
+        | Some v -> Some v
+      end
+    end
 
-  and 'a local_state =
-    { solver : solver
-    ; mutable next : hold option
-    ; global : 'a global_state
-    }
+  let get_model_or_stop symbol =
+    let* model = get_model symbol in
+    match model with Some v -> return v | None -> stop
 
-  and e_local_state = E_st : 'a local_state -> e_local_state [@@unboxed]
+  let select (cond : S.vbool) =
+    let v = Expr.simplify cond in
+    match v.node.e with
+    | Val True -> return true
+    | Val False -> return false
+    | Val (Num (I32 _)) -> failwith "unreachable (type error)"
+    | _ ->
+      let true_branch =
+        let* () = add_pc v in
+        let* () = check_reachability in
+        return true
+      in
+      let false_branch =
+        let* () = add_pc (S.Bool.not v) in
+        let* () = check_reachability in
+        return false
+      in
+      choose true_branch false_branch
+  [@@inline]
+
+  let summary_symbol (e : Expr.t) =
+    let* thread in
+    match e.node.e with
+    | Symbol sym -> return (None, sym)
+    | _ ->
+      let choices = thread.choices in
+      let symbol_name = Format.sprintf "choice_i32_%i" choices in
+      let+ () = modify_thread (fun t -> { t with choices = choices + 1 }) in
+      let sym = Symbol.(symbol_name @: Ty_bitv S32) in
+      let assign = Expr.(Relop (Eq, mk_symbol sym, e) @: Ty_bitv S32) in
+      (Some assign, sym)
+
+  let select_i32 (i : S.int32) =
+    let sym_int = Expr.simplify i in
+    match sym_int.node.e with
+    | Val (Num (I32 i)) -> return i
+    | _ ->
+      let* assign, symbol = summary_symbol sym_int in
+      let* () =
+        match assign with Some assign -> add_pc assign | None -> return ()
+      in
+      let rec generator () =
+        let* possible_value = get_model_or_stop symbol in
+        let i =
+          match possible_value with
+          | Num (I32 i) -> i
+          | _ -> failwith "Unreachable: found sumbol must be a value"
+        in
+        let this_value_cond = Expr.Bitv.I32.(Expr.mk_symbol symbol = v i) in
+        let not_this_value_cond =
+          Expr.Bitv.I32.(Expr.mk_symbol symbol != v i)
```

Comment:
`!=` is the negation of `==` (physical equality), you probably want `<>` instead ?

Reply:
It's actually Expr.Vitv.I32's `!=` : https://github.com/formalsec/encoding/blob/main/lib/expr.ml#L359